### ISSUE TITLE
Expose code bundling capabilites in Wing to be used in downstream libraries / projects

### ISSUE THREAD

**Issue by skorfmann** [winglang/wing#5516]
### Use Case

in https://github.com/winglang/winglibs/pull/35#issuecomment-1903522671 synthesized inflight code is in conjunction with a shell provider https://registry.terraform.io/providers/scottwinkler/shell/latest/docs The goal there is to define an ad-hoc, dynamic Terraform provider resource in Wing itself. The inflight code will run as part of the Terraform apply on the machine Terraform is being executed.

See the wing code https://github.com/winglang/winglibs/pull/35/files#diff-5814e06c93e84a8fced40419c598b55392de082a05007b0551aa3a7cfefc7ec3

Quoting the comment around the code which used to make it work by @eladb 

>    // TODO: this is a hack calling the internal API. I am wondering if we should expose this
>    // capability through some other utility API (`std.toInflightJavaScript(handler)`?). eventually
>    // this returns some leaky JavaScript code (for inflight closures it's a class that has a
>    // `handle` method), so I am not sure how nice we can make this.

with https://github.com/winglang/wing/pull/5306 this stopped working, since there's now a global helper required which needs the `@winglang/sdk` npm package available. 

### Proposed Solution

When thinking about this, the `tfresource` winglib should bundle the handler code as it would be done for Lambda handlers. So perhaps the bundling capabilities could be exposed in wing itself? Or maybe it's possible already?

### Component

Compiler, SDK, Libraries

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.
* If this issue is labeled **needs-discussion**, it means the spec has not been finalized yet. Please reach out on the #dev channel in the [Wing Slack](https://t.winglang.io/slack).

**Comment by eladb** [comment]
I support this message, @skorfmann can you propose a solution?

**Comment by Chriscbr** [comment]
Yeah, this is a tricky issue - but it's important since converting inflight closures into JavaScript is necessary for anyone else to design classes for custom compute resources in Wing.

It's worth noting that calling `toInflight()` in the manner `tfresource` library currently does should be invalid because it's called before all of the construct tree has finished resolving. This can result in inflight code that is invalid in subtle ways.

In https://github.com/winglang/wing/pull/5453, to fix a number of bugs relating to laziness, we reworked the code in our current compute resources (`cloud.Function` and `cloud.Service`) so that they only call `toInflight()` during an internal `preSynthesis` step. Ideally perhaps we'd like to remodel this in some way (with a public API) that avoids exposing users to this `preSynthesis` step.

We also need to make a distinction between APIs that expect/return unbundled code (like the current hidden `_toInflight()` method) and interfaces that expect/return bundled code. In principle we could always bundle code and eliminate that distinction, but for performance reasons I don't think it's viable right now.
